### ISSUE TITLE
Ft's @Timeout does not interrupt/terminate rest calls

### ISSUE THREAD

**Issue by pilhuhn** [eclipse/microprofile-rest-client#107]
I would assume that in the following scenario `getMessage()` is aborted with an exception if `service.getSlow()` takes more than 200ms, but this currently does not happen.

```
  @Inject
  @RestClient
  BackendService service;

 @Timeout(200)
  String getMessage(String mode) {
     Response r = service.getSlow();
     [..]
  }
```

cc @Emily-Jiang

**Comment by kenfinnigan** [comment]
I don't believe REST Client currently supports FT integration

**Comment by Emily-Jiang** [comment]
@kenfinnigan It is not quite about REST client integration. The issue is that the JAX RS worker is not interruptible. @pilhuhn @Azquelt hit this issue as well. @andymc12 I think  @Azquelt  discussed this with you briefly at EclipseCon France. Did you run this past cxf community?

**Comment by andymc12** [comment]
Hi @Emily-Jiang , yes, I spoke with @Azquelt about this at EclipseCon.  As you mentioned, establishing an HTTP connection and then reading from it are blocking operations.  Interrupting the thread has no effect.  Andrew and I discussed the possibility of obtaining the timeout value from a thread local and then dynamically adjusting the JAX-RS/MP Rest Client's connection and read timeouts accordingly such that the `@Timeout` annotation could be honored.

**Comment by andymc12** [comment]
Finally getting back to this.  

I some testing and prototyping (based on CXF) that might help us here.  For asynchronous requests (either when using FT's `@Asynchronous` or Rest Client's async methods (returning `CompletionStage`), the timeout was never a factor because the request was handled on a separate thread, so even if the request took longer than the value specified in the `@Timeout` annotation, the method was able to return normally, and then the caller would just have to deal with either a `Future.get(...)` or similar with the `CompletionStage`.  I think this is working as expecting.

So, I think this issue really only comes into play for synchronous client requests.  In my prototype, I overrode the request's connect and read timeout with the value specified in `@Timeout` plus 2 milliseconds.  This had the effect of timing out very close to the requested timeout, but allowing FT to throw it's `TimeoutException` rather than an implementation-specific version of `SocketException` or `IOException`.  (By adding the 2 milliseconds, it guarantees that the connect or read timeout will happen after FT's timer expires - that's why the `TimeoutException` gets thrown.)

For the vast majority of cases, I think this should be sufficient.  The one case where it won't is when the rest client is making an initial (or unpooled) request to a remote service that is slow to connect and slow to respond.  For example, suppose the user set `@Timeout(2000)` (2000 milliseconds) and the remote service takes 1900 ms to connect and then another 1900 ms to respond.  If the socket read is not interruptible, then the timeout may not fire until the request completes - after 3800 ms.  

I'm leaning toward using this approach in the spec document, but also documenting this potential issue.  I think the majority of use cases will be pooled (pre-connected) connections and/or non-blocking I/O on reading the response. 

Does that sound ok?

**Comment by andymc12** [comment]
closed with PR #140
